autometa/binning/recursive_dbscan.py

```diff
@@ -180,95 +281,131 @@ def run_dbscan(df, eps, dropcols=['cluster', 'purity', 'completeness'],
+    clusterer = HDBSCAN(
+        # cluster_selection_epsilon=eps,
+        min_cluster_size=min_cluster_size,
+        min_samples=min_samples,
+        cluster_selection_method="leaf",
+        allow_single_cluster=True,
+    ).fit(X)
+    clusters = pd.Series(clusterer.labels_, index=df.index, name="cluster")
+    return pd.merge(df, clusters, how="left", left_index=True, right_index=True)
+
+
+def recursive_hdbscan(
+    table, markers_df, domain, completeness_cutoff, purity_cutoff, verbose=False,
+):
+    """Recursively run HDBSCAN starting with defaults and iterating the min_samples
+     and min_cluster_size until only 1 cluster is recovered.
 
     Parameters
     ----------
-    df : pd.DataFrame
-        index='contig' cols=['x','y','coverage','cluster']
+    table : pd.DataFrame
+        Contigs with embedded k-mer frequencies as ['x','y','z'] columns and
+        optionally 'coverage' column
     markers_df : pd.DataFrame
         wide format, i.e. index=contig cols=[marker,marker,...]
-    domain : str, optional
+    domain : str
         Kingdom to determine metrics (the default is 'bacteria').
         choices=['bacteria','archaea']
+    completeness_cutoff : float
+        `completeness_cutoff` threshold to retain cluster (the default is 20.0).
+    purity_cutoff : float
+        `purity_cutoff` threshold to retain cluster (the default is 90.0).
+    verbose : bool
+        log stats for each recursive_dbscan clustering iteration.
 
     Returns
     -------
     2-tuple
-        `df` with added columns - 'completeness' and 'purity'
-        pd.DataFrame of clusters with their 'completeness' and 'purity'
-
-    Raises
-    -------
-    KeyError
-        `domain` is not "bacteria" or "archaea"
+        (pd.DataFrame(<passed cutoffs>), pd.DataFrame(<failed cutoffs>))
+        DataFrames consisting of contigs that passed/failed clustering
+        cutoffs, respectively.
 
+        DataFrame:
+            index = contig
+            columns = ['x,'y','z','coverage','cluster','purity','completeness']
     """
-    domain = domain.lower()
-    marker_sets = {'bacteria': 139.0, 'archaea': 162.0}
-    if domain not in marker_sets:
-        raise KeyError(f'{domain} is not bacteria or archaea!')
-    expected_number = marker_sets[domain]
-    clusters = dict(list(df.groupby('cluster')))
-    metrics = {'purity': {}, 'completeness': {}}
-    for cluster, dff in clusters.items():
-        contigs = dff.index.tolist()
-        summed_markers = markers_df[markers_df.index.isin(contigs)].sum()
-        is_present = summed_markers >= 1
-        is_single_copy = summed_markers == 1
-        nunique_markers = summed_markers[is_present].index.nunique()
-        num_single_copy_markers = summed_markers[is_single_copy].index.nunique(
-        )
-        completeness = nunique_markers / expected_number * 100
-        # Protect from divide by zero
-        if nunique_markers == 0:
-            purity = pd.NA
+
+    min_cluster_size = 2
```

Comment:
Can we start with `min_cluster_size = 1`? This would be equivalent to the `min_samples = 1` parameter we use for DBSCAN. The reason why I chose that is that there can be single contigs which legitimately are also a whole bin. For example AB1_lowGC that we found in the first *Bugula* metagenome.

Reply:
You should probably test this on AB1, because I am wondering whether this method of sweeping the parameter with HDBSCAN would mean that AB1_lowGC is always grouped with other things.

Reply:
This results in a ValueError from hdbscan:


<details open><summary><code>(autometa) ➜  Autometa git:(WiscEvan/issue52) ✗ python -m autometa.binning.recursive_dbscan --clustering-method hdbscan --taxonomy tests/dtc_data/taxonomy.tsv --embedded-kmers tests/dtc_data/kmers.bhsne.tsv tests/dtc_data/kmers.tsv tests/dtc_data/coverages.tsv tests/dtc_data/bacteria.markers.tsv tests/dtc_data/binning.tsv</code></summary><p>
<!-- between the ticks below, paste the output of 'autometa-configure --dryrun --debug' -->

```
06/18/2020 01:04:43 PM : autometa.common.kmers : DEBUG : k-mers frequency embedding already exists tests/dtc_data/kmers.bhsne.tsv
06/18/2020 01:04:43 PM : root : DEBUG : master_df shape: (2829, 11)
06/18/2020 01:04:43 PM : __main__ : INFO : Examining superkingdom: 1 unique taxa (2,829 contigs)
06/18/2020 01:04:43 PM : __main__ : DEBUG : Examining taxonomy: superkingdom : bacteria : (2829, 11)
06/18/2020 01:04:43 PM : __main__ : INFO : clustering with recursive_hdbscan
Traceback (most recent call last):
  File "/Users/rees/miniconda3/envs/autometa/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Users/rees/miniconda3/envs/autometa/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 710, in <module>
    main()
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 701, in main
    verbose=args.verbose,
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 594, in binning
    verbose=verbose,
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 446, in get_clusters
    master_df, markers_df, domain, completeness, purity, verbose=verbose,
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 352, in recursive_hdbscan
    binned_df = run_hdbscan(table, min_cluster_size)
  File "/Users/rees/Autometa/autometa/binning/recursive_dbscan.py", line 305, in run_hdbscan
    ).fit(X)
  File "/Users/rees/miniconda3/envs/autometa/lib/python3.7/site-packages/hdbscan/hdbscan_.py", line 919, in fit
    self._min_spanning_tree) = hdbscan(X, **kwargs)
  File "/Users/rees/miniconda3/envs/autometa/lib/python3.7/site-packages/hdbscan/hdbscan_.py", line 494, in hdbscan
    raise ValueError('Min cluster size must be greater than one')
ValueError: Min cluster size must be greater than one
```
</p></details>